doc/tutorial/tutorial.md

```diff
@@ -1,40 +1,650 @@
+MCastAddress(a) :- Address(a), is_multicast_addr(a).
+```
+
+DDlog offers two ways to write bit vector literals. First, they can be written as decimal numbers.
+DDlog performs *type inference* to determine whether the number should be interpreted as an unbounded
+mathematical integer or a bit vector.  In the latter case it also infers its bit width.  For
+example, the literal `125` is interpreted as a mathematical integer in `var x: int = 125` and as an
+8-bit bit-vector in `var x: bit<8> = 125`.
+
+Second, bit vector literals can be written with explicit width and radix specifiers using the
+`[<width>]<radix_specifier><value>` syntax, where `<radix>` is one of `'d`, `'h`, `'o`, `'b` for
+decimal, hexadecimal, octal and binary numbers respectively:
+
+```
+32'd15              // 32-bit decimal
+48'haabbccddeeff    // 48-bit hexadecimal
+3'b101              // 3-bit binary
+'d15                // bit vector whose value is 15 and whose width is determined from the context
+```
 
 ### Control flow and variables
 
-*if-then-else, sequencing, assignments, var*
-
-## Procedures
-
-### Extern procedures
+Complex computations can be expressed in DDlog using control-flow constructs:
+
+1. *Sequencing* evaluates semicolumn-separated expressions in order.
+
+1. *`if (cond) e1 else e2`* evaluates one of its subexpressions depending on the value of `cond`.
+
+1. Matching chooses among one of several expressions by matching its argument against a series of
+patterns.
+
+Note that DDlog currently does not support loops.
+
+The following example illustrates these constructs:
+
+```
+function addr_port(ip: ip_addr_t, proto: string, preferred_port: bit<16>): string =
+{
+    var port: bit<16> = match (proto) {
+        "FTP"   -> 20,  // default FTP port
+        "HTTPS" -> 443, // default HTTP port
+        _       -> {    // other protocol
+            if (preferred_port != 0)
+                preferred_port // return preferred_port if specified
+            else
+                16'd80         // assume HTTP otherwise
+        }
+    };
+    // Return the address:port string
+    $"${ip}:${port}"
+}
+```
+
+This example highlights several important aspects of DDlog's control flow constructs.  First, note
+that the function does not have a `return` statement (in fact, there is currently no `return`
+statemnt in DDlog).  DDlog is an *expression-oriented language*, which does not differentiate
+between expressions and statements.  All control flow constructs are expressions that have a
+statically assigned type and a dynamically assigned value and can be used as terms in other
+expressions.  The function's return value is simply the value produced by the expression in its
+body.  In particular, a sequential block returns the value of the last expression in the sequence;
+`if-else` and `match` expressions return the value of the taken branch.  As a consequence, the
+`else` clause is not optional in an `if-else` expression, and match clauses must be exhaustive.
```

Comment:
In a match clause there should always exist a matching pattern - you can give perhaps a counterexample to be clear.